x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java

```diff
@@ -279,6 +284,79 @@ private static void checkAggregate(LogicalPlan p, Set<Failure> failures) {
         }
     }
 
+    /**
+     * Check CATEGORIZE grouping function usages.
+     * <p>
+     *     Some of those checks are temporary, until the required syntax or engine changes are implemented.
+     * </p>
+     */
+    private static void checkCategorizeGrouping(Aggregate agg, Set<Failure> failures) {
+        // Forbid CATEGORIZE grouping function with other groupings
+        if (agg.groupings().size() > 1) {
+            agg.groupings().forEach(g -> {
+                g.forEachDown(
+                    Categorize.class,
+                    categorize -> failures.add(
+                        fail(categorize, "cannot use CATEGORIZE grouping function [{}] with multiple groupings", categorize.sourceText())
+                    )
+                );
+            });
+        }
+
+        // Forbid CATEGORIZE grouping function within other functions
+        agg.groupings().forEach(g -> {
+            g.forEachDown(
+                Function.class,
+                function -> function.children()
+                    .forEach(
+                        child -> child.forEachDown(
+                            Categorize.class,
+                            categorize -> failures.add(
+                                fail(
+                                    categorize,
+                                    "cannot use CATEGORIZE grouping function [{}] within another function",
+                                    categorize.sourceText()
+                                )
+                            )
+                        )
+                    )
+            );
+        });
+
+        // Forbid CATEGORIZE being used in the aggregations
+        agg.aggregates().forEach(a -> {
+            a.forEachDown(
+                Categorize.class,
+                categorize -> failures.add(
+                    fail(categorize, "cannot use CATEGORIZE grouping function [{}] within the aggregations", categorize.sourceText())
+                )
+            );
```

Comment:
This may be slightly overconstraining, resp. inconsistent with `BUCKET`. The following works on main:
`| stats mv_count(bucket(x, 1)) by bucket(x, 1)`

Reply:
Correction: Such special condition isn't for BUCKET, but for all GroupingFunctions. I'll recheck it tomorrow and see if it should, indeed, work!

Reply:
Agree! Consistency with `BUCKET` is not important for now, and we can still allow this later if needed.